lib/pure/net.nim

```diff
@@ -120,21 +120,16 @@ when defineSsl:
     SslContext* = ref object
       context*: SslCtx
       referencedData: HashSet[int]
-      extraInternal: SslContextExtraInternal
 
     SslAcceptResult* = enum
       AcceptNoClient = 0, AcceptNoHandshake, AcceptSuccess
 
     SslHandshakeType* = enum
       handshakeAsClient, handshakeAsServer
 
-    SslClientGetPskFunc* = proc(hint: string): tuple[identity: string, psk: string]
+    SslClientGetPskFunc* = proc(hint: string): tuple[identity: string, psk: string]{.nimcall.}
 
-    SslServerGetPskFunc* = proc(identity: string): string
-
-    SslContextExtraInternal = ref object of RootRef
-      serverGetPskFunc: SslServerGetPskFunc
-      clientGetPskFunc: SslClientGetPskFunc
+    SslServerGetPskFunc* = proc(identity: string): string{.nimcall.}
```

Comment:
This is a breaking change. The only reason to justify this change is that it's security critical. Is that the case?

Reply:
status quo is: servers using the psk API in net, in Nim >0.15.2 will compile and run happily, but sigsegv when a client initiates a connection with psk.

this is a DDoS vulnerability, at least.

Lets say Hyrum's Law applies, and someone is using this to reliably crash their servers remotely, and have set it up to work with a closure. That will no longer compile. I understand that's a break.

Maybe there's a way to allow embedding a closure in a {.cdecl.} callback but at the risk of sounding lazy or stupid I 
a) couldn't figure it out:
  the farthest i can get is `tnetpsk.nim(42, 30) Error: internal error: inconsistent environment type`
b) I doubt the benefit of the added code complexity.  ref your first comment on this pr: https://github.com/nim-lang/Nim/pull/17741#discussion_r615233471

no ego here, close away if my solution isn't aligned with Nim's requirements, but i honestly can't figure out how to make this work any other way.  The fact that I had to add an `openssl` dependency to the azure osx CI just for this test alone makes this whole PR a bit suspect.

Reply:
nvm I've an idea

Reply:
this should work: https://github.com/nim-lang/Nim/pull/17741#issuecomment-825933893